lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        {
+                            _systemState.MoveNext(SystemStateEvent.AllTasksAreRunning);
+                            Logger.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "All tasks are running, SystemState {0}", _systemState.CurrentState));
+                        }
+                        break;
+                    case SystemState.ShuttingDown:
+                    case SystemState.Fail:
+                        _taskManager.RecordRunningTaskDuringSystemFailure(runningTask, TaskManager.CloseTaskByDriver);
+                        break;
+                    default:
+                        UnexpectedState(runningTask.Id, "IRunningTask");
+                        break;
+                }
             }
-            Exceptions.Throw(new Exception(string.Format("Data Loading Context with Id: {0} failed", value.Id)), Logger);
         }
+        #endregion IRunningTask
 
+        #region ICompletedTask
         /// <summary>
-        /// Specifies what to do if a task fails.
-        /// We throw the exception and fail IMRU unless IMRU job is already done.
+        /// ICompletedTask handler. It is called when a task is completed. The following action will be taken based on the System State:
+        /// Case TasksRunning
+        ///     Updates task state to TaskCompleted
+        ///     If all tasks are completed, sets system state to TasksCompleted and then go to Done action
+        /// Case ShuttingDown
+        ///     Updates task state to TaskCompleted
+        ///     Try to recover
+        /// Other cases - not expected 
         /// </summary>
-        /// <param name="value"></param>
-        public void OnNext(IFailedTask value)
+        /// <param name="completedTask">The link to the completed task</param>
+        public void OnNext(ICompletedTask completedTask)
         {
-            if (AreIMRUTasksCompleted())
+            Logger.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "Received ICompletedTask {0}, systemState {1} in retry # {2}.", completedTask.Id, _systemState.CurrentState, _numberOfRetryForFaultTolerant));
+
+            lock (_lock)
             {
-                Logger.Log(Level.Info,
-                    string.Format("Task with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
-                return;
+                switch (_systemState.CurrentState)
+                {
+                    case SystemState.TasksRunning:
+                        _taskManager.RecordCompletedTask(completedTask);
+                        if (_taskManager.AreAllTasksCompleted())
+                        {
+                            _systemState.MoveNext(SystemStateEvent.AllTasksAreCompleted);
+                            Logger.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "All tasks are completed, systemState {0}", _systemState.CurrentState));
+                            DoneAction();
```

Comment:
This method is to dispose all the contexts. We would like to get all the task completed and then dispose contexts together.